Source/CesiumRuntime/Private/CesiumDmsEditor.cpp

```diff
@@ -0,0 +1,290 @@
+// Copyright 2020-2021 CesiumGS, Inc. and Contributors
+
+#if WITH_EDITOR
+
+#include "CesiumDmsEditor.h"
+#include "CesiumRuntime.h"
+
+#include "PropertyCustomizationHelpers.h"
+#include "PropertyEditing.h"
+
+#include "Widgets/Input/SSpinBox.h"
+#include "Widgets/Input/STextComboBox.h"
+#include "Widgets/Text/STextBlock.h"
+
+#include <glm/glm.hpp>
+
+/**
+ * @brief A structure describing cartographic coordinates in
+ * the DMS (Degree-Minute-Second) representation.
+ */
+struct DMS {
+
+  /**
+   * @brief The degrees.
+   *
+   * This is usually a value in [0,90] (for latitude) or
+   * in [0,180] (for longitude), although explicit
+   * clamping is not guaranteed.
+   */
+  int32_t d;
+
+  /**
+   * @brief The minutes.
+   *
+   * This is a value in [0,60).
+   */
+  int32_t m;
+
+  /**
+   * @brief The seconds.
+   *
+   * This is a value in [0,60).
+   */
+  double s;
+
+  /**
+   * @brief Whether the coordinate is negative.
+   *
+   * When the coordinate is negative, it represents a latitude south
+   * of the equator, or a longitude west of the prime meridian.
+   */
+  bool negative;
+};
+
+/**
+ * @brief Converts the given decimal degrees to a DMS representation.
+ *
+ * @param decimalDegrees The decimal degrees
+ * @return The DMS representation.
+ */
+DMS decimalDegreesToDms(double decimalDegrees) {
+  // Roughly based on
+  // https://en.wikiversity.org/wiki/Geographic_coordinate_conversion
+  // Section "Conversion_from_Decimal_Degree_to_DMS"
+  bool negative = decimalDegrees < 0;
+  double dd = negative ? -decimalDegrees : decimalDegrees;
+  double d = glm::floor(dd);
+  double min = (dd - d) * 60;
+  double m = glm::floor(min);
+  double sec = (min - m) * 60;
+  double s = sec;
+  if (s >= 60) {
+    m++;
+    s -= 60;
+  }
+  if (m == 60) {
+    d++;
+    m = 0;
+  }
+  return DMS{static_cast<int32_t>(d), static_cast<int32_t>(m), s, negative};
+}
+
+/**
+ * @brief Converts the given DMS into decimal degrees.
+ *
+ * @param dms The DMS
+ * @return The decimal degrees
+ */
+double dmsToDecimalDegrees(const DMS& dms) {
+  double dd = dms.d + (dms.m / 60.0) + (dms.s / 3600.0);
+  if (dms.negative) {
+    return -dd;
+  }
+  return dd;
+}
+
+CesiumDmsEditor::CesiumDmsEditor(
```

Comment:
The abbreviation "dms" is ok for the private `dmsToDecimalDegrees`, and I can even live with the `DMS` struct (but should both of those be in the anonymous namespace?), but for a file and class name, `CesiumDmsEditor` really should be `CesiumDegreesMinutesSecondsEditor` even though it's a mouthful. Though we sometimes bend it, the official Cesium rule is "no abbreviations or acronyms at all."